[Data] Extend API to enable passing sample weights via ray.dataset.to_tf (#45701)

jeffreyjeffreywang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
`tensorflow.keras.model.fit` supports callers to pass in sample weights along with input dataset (https://www.tensorflow.org/versions/r2.8/api_docs/python/tf/keras/Model#fit). However, `ray.data.Dataset.to_tf()` does not support this feature. Currently, callers can only generate data instances comprised of features and labels.

This PR extends `ray.data.Dataset.to_tf()` to allow callers to specify additional metadata associated with data samples. Specifically, `additional_columns` is introduced as a forced kwargs to `ray.data.Dataset.to_tf()` and `ray.DataIterator.to_tf()`. When `additional_columns` is not specified, there is no change to the API behavior. By contrast, while `additional_columns` is provided, the APIs will yield `additional_metadata` along with `features` and `labels`. To leverage the functionality for passing sample weights along with sample features and labels in tensorflow, one can create `tf.data.Dataset` from `ray.data.Dataset.to_tf()` and specifying `additional_columns="weight"` as `weight` is the column for storing sample weights in `ray.data.Dataset`.

We do not explicitly name the new argument `sample_weight` or `weight` and do not limit its type to be `str` as there may be other metadata associated with each sample that we want to yield while iterating through the dataset.

We follow the heuristic of inferring `feature_type_spec` and `label_type_spec` when determining `additional_type_spec`. Existing tests are extended to validate when the APIs are invoked with and without the `additional_columns` argument to ensure existing callers won't fail to invoke `ray.data.Dataset.to_tf()` and `ray.DataIterator.to_tf()` as this new change rolls out.

## Related issue number
Resolves #45701 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jeffreyjeffreywang commented on python/ray/data/dataset.py:
```diff
@@ -4059,13 +4059,15 @@ def to_tf(
         feature_columns: Union[str, List[str]],
         label_columns: Union[str, List[str]],
         *,
+        additional_columns: Union[str, List[str]] = None,
```
I'm not thrilled about the naming of this argument. Open for brainstorming 😄 

Even though the current use case for this argument is to include sample weights, I don't think naming it as `sample_weight` or `weight` is a good idea as there may be other metadata that we want incorporate/yield in the future.

jeffreyjeffreywang replied:
I also deliberately allow `List[str]` as its type as we may want to associate more than one metadata to each (feature, label) pair.

jeffreyjeffreywang commented on python/ray/data/iterator.py:
```diff
@@ -676,13 +676,15 @@ def to_tf(
+        additional_type_spec: Union["tf.TypeSpec", Dict[str, "tf.TypeSpec"]] = None,
     ) -> "tf.data.Dataset":
         """Return a TF Dataset over this dataset.
 
```
Hey @scottjlee, let me know if you think adding an example specifically for explaining the usage of the new argument is a good idea.

scottjlee replied:
yes good idea, could you add an example that demonstrates how to use `feature_type_spec`, `label_type_spec`, and `additional_type_spec`? thanks!

jeffreyjeffreywang replied:
Yeah definitely, will do!

jeffreyjeffreywang commented on python/ray/data/tests/test_tf.py:
```diff
@@ -28,76 +28,179 @@ def test_autosharding_is_disabled(self):
-        ds = ray.data.from_items([{"spam": 0, "ham": 0}])
-
-        dataset = ds.to_tf(feature_columns="spam", label_columns="ham")
+    @pytest.mark.parametrize("include_additional_columns", [False, True])
```
Note for reviewers: I intend to validate the API with and without (which shall fall back to the original behavior) the new arguments to ensure that the API is backward compatible. I understand there may be some redundancy in the following tests. Let me know if you think there is a better approach for testing.

scottjlee reviewed (commented):
thanks for updating all the tests!

can we also add an additional unit test, where we use `Dataset.to_tf()` with sample weights passed, then we call `keras.Model.fit()` with the sample weights? so that we can verify that the original intended behavior

jeffreyjeffreywang commented on python/ray/data/iterator.py:
```diff
@@ -842,9 +861,25 @@ def generator():
+            validate_columns(additional_columns)
+            additional_type_spec = get_type_spec(schema, columns=additional_columns)
+
+        if additional_columns is None:
+            dataset = tf.data.Dataset.from_generator(
+                generator, output_signature=(feature_type_spec, label_type_spec)
+            )
+        else:
+            dataset = tf.data.Dataset.from_generator(
+                generator,
+                output_signature=(
+                    feature_type_spec,
+                    label_type_spec,
+                    additional_type_spec,
+                ),
+            )
```
Hey @scottjlee, thank you so much for taking some time to go over my PR! I'll address your comments and add the test you mentioned above.

jeffreyjeffreywang replied:
The test for using `Dataset.to_tf()` with sample weights passed and calling `keras.Model.fit()` with the sample weights is actually already included with TestToTF.test_training with `include_additional_columns=True`. @scottjlee , let me know if you think this test is sufficient to validate the original intended behavior.

scottjlee replied:
ah yeah i missed that, thanks! that should be sufficient.

jeffreyjeffreywang commented on python/ray/data/iterator.py:
```diff
@@ -842,9 +861,25 @@ def generator():
-        )
+        if additional_columns is not None and additional_type_spec is None:
+            schema = self.schema()
+            valid_columns = set(schema.names)
```
`valid_columns` is used in `validate_columns()` to verify `additional_columns` refers to an existing column name. If we remove this line and `feature_type_spec` and `label_type_spec` are both not `None`, `valid_columns` will not be set on line 858. When we enter this `if` block on line 864, there will be no `valid_columns` to check against.

scottjlee replied:
yes you're right, my mistake

jeffreyjeffreywang commented:
@scottjlee Would appreciate your quick review when you have time, no rush 😄

scottjlee reviewed (approved):
thanks for the contribution!

omatthew98 reviewed (approved):
Lgtm, thanks for the contribution!

scottjlee commented:
@jeffreyjeffreywang let me know when the docs example is added, and i can take a final look before we get this merged

jeffreyjeffreywang commented:
Hey @scottjlee, definitely, I plan to get it done today.

jeffreyjeffreywang commented:
Hey @scottjlee, I've added some documentation to explain the spec arguments and the new argument I introduced. Doc examples are validated with `pytest --dodctest-modules` and built locally with `make develop` in the doc directory. Here are the sections/examples I added. Please let me know if these look good to you 😄 

### ray.data.Dataset.to_tf
<img width="884" alt="Screenshot 2024-08-01 at 6 00 37 PM" src="https://github.com/user-attachments/assets/44d60414-e486-4efc-8797-b56cf97842ef">

### ray.data.DataIterator.to_tf
<img width="881" alt="Screenshot 2024-08-01 at 6 01 59 PM" src="https://github.com/user-attachments/assets/2bf27094-8f67-4ddd-9bf4-28df2a05cf83">

scottjlee commented:
Thanks, the docs changes look good to me. I am merging in latest master to resolve the build issue (unrelated to this PR)

jeffreyjeffreywang commented:
Thanks Scott! Looking forward to contributing more 😃
